custom_enumerable_methods.rb

```diff
@@ -0,0 +1,261 @@
+    return false unless block_given? || !my_parameter.nil?
+
+    result = true
+    if self.class == Array
+      my_each do |x|
+        if block_given?
+          result = false unless yield(x)
+        elsif my_parameter.class == Regexp
+          result = false unless x.match(my_parameter)
+        elsif my_parameter.class <= Numeric
+          result = false unless x == my_parameter
+        else
+          result = false unless x.class <= my_parameter
+        end
+        break unless result
+      end
+    else
+      my_each do |key, value|
+        result = false unless yield(key, value)
+      end
+    end
+    result
+  end
+
+  def my_any?(our_parameter = nil)
+    return false if (self.class == Array && count.zero?) || (!block_given? &&
+    our_parameter.nil? && !include?(true))
+    return true unless block_given? || !our_parameter.nil?
+
+    result = false
+    if self.class == Array
+      my_each do |i|
+        if block_given?
+          result = true if yield(i)
+        elsif our_parameter.class == Regexp
+          result = true if i.match(our_parameter)
+        elsif our_parameter.class == String
+          result = true if i == our_parameter
+        elsif i.class <= our_parameter
+          result = true
+        end
+      end
+    else
+      my_each do |key, value|
+        result = true if yield(key, value)
+      end
+    end
+    result
+  end
+
+  def my_none?(my_parameter = nil)
+    return true if count.zero? || (self[0].nil? && !include?(true))
+    return false unless block_given? || !my_parameter.nil?
+
+    result = true
+    if self.class == Array
+      my_each do |x|
+        if block_given?
+          result = false if yield(x)
+        elsif my_parameter.class == Regexp
+          result = false if x.match(my_parameter)
+        elsif my_parameter.class <= Numeric
+          result = false if x == my_parameter
+        elsif x.class <= my_parameter
+          result = false
+        end
+        break unless result
+      end
+    else
+      my_each do |key, value|
+        result = false if yield(key, value)
+        break unless result
+      end
+    end
+    result
+  end
+
+  def my_count(my_variable = nil)
+    my_counter = 0
+    if block_given?
+
+      if self.class == Array
+        my_each do |n|
+          my_counter += 1 if yield(n)
+        end
+      else
+        my_each do |key, value|
+          my_counter += 1 if yield(key, value)
+        end
+      end
+    elsif !block_given? && my_variable.nil?
+      return length
+    elsif !block_given? && !my_variable.nil?
+      my_each do |n|
+        my_counter += 1 if n == my_variable
+      end
+    end
+    my_counter
+  end
+
+  def my_map
+    return to_enum unless block_given?
+
+    new_array = []
+    if self.class == Array
+      my_each do |n|
+        new_array << yield(n)
+      end
+    else
+      my_each do |key, value|
+        new_array << yield(key, value)
+      end
+    end
+    new_array
+  end
```

Comment:
`my_map` identical and returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://github.com/TheOdinProject/curriculum/blob/master/ruby_programming/archive/basic_ruby/project_advanced_building_blocks.md#project-2-enumerable-methods))
![image](https://user-images.githubusercontent.com/38625956/82838260-72e0e300-9eed-11ea-9c3e-46a340a1d2fc.png)
- executes only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum.
```ruby
my_proc = Proc.new { |num| num > 10 }
[11, 2, 3, 15].my_map(my_proc) {|num| num < 10 } # should return [true, false, false, true]
```